New DeploymentStrategy type for JaegerSpec.Strategy

volmedo opened the pull request:
This PR adds a new type, `DeploymentStrategy`, to be used in `JaegerSpec.Strategy` instead of just a `string`. Constants based on the new type are also defined to define valid values.

When merged, this PR will close #137. 

### To do
- [x] make sure all unserialized values are converted to lower case
- [x] a test ensuring that current CRs with a mixed-case strategy are correctly parsed
- [x] the constant in the strategy package, mentioned in #137

jpkrohling commented on pkg/deployment/collector.go:
```diff
@@ -76,9 +75,9 @@ func (c *Collector) Get() *appsv1.Deployment {
+	// If strategy is DeploymentStrategyStreaming, then change storage type
 	// to Kafka, and the storage options will be used in the Ingester instead
-	if strings.EqualFold(c.jaeger.Spec.Strategy, "streaming") {
+	if c.jaeger.Spec.Strategy == v1.DeploymentStrategyStreaming {
```
What would happen if we have a CR with a strategy set to `Streaming`?

volmedo replied:
I was assuming that the value was being normalized somewhere else. Sorry, my bad.

jpkrohling commented on pkg/apis/jaegertracing/v1/jaeger_types.go:
```diff
@@ -49,11 +49,33 @@ const (
 
+// DeploymentStrategy represents the possible values for deployment strategies
+// +k8s:openapi-gen=true
+type DeploymentStrategy string
```
You could implement the UnmarshalJSON method to the type, to ensure the values read from JSON are converted to lowercase.

volmedo commented:
> 1. make sure all unserialized values are converted to lower case

Where is the unmarshalling being done? I suppose it is done by some library from k8s or the operator framework. If that is the case, is it enough to make the new type implement the `Unmarshaler` interface?

> 2. possibly a test ensuring that current CRs with a mixed-case strategy are correctly parsed

Where should this test live? Are there any other similar tests?

jpkrohling commented:
> Where is the unmarshalling being done?

It's done by the Operator SDK. It's sufficient to have a couple of methods, like these from the `Options` object:

https://github.com/jaegertracing/jaeger-operator/blob/d11973f0dbe9b7b602da1f1214e21d40c4a5e8dd/pkg/apis/jaegertracing/v1/options.go#L41-L62

I *think* you only need to implement the `UnmarshalJSON`, but needs confirmation. 

We have some tests for the same object that could be used as reference:

https://github.com/jaegertracing/jaeger-operator/blob/d11973f0dbe9b7b602da1f1214e21d40c4a5e8dd/pkg/apis/jaegertracing/v1/options_test.go#L11-L16

jpkrohling commented on pkg/strategy/strategy.go:
```diff
@@ -9,12 +9,13 @@ import (
 	"k8s.io/api/extensions/v1beta1"
 	rbac "k8s.io/api/rbac/v1"
 
+	jv1 "github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1"
```
We typically use `v1` for Jaeger, and `corev1` for `k8s.io/api/core/v1`

volmedo replied:
Makes a lot of sense!

I saw that `v1` was already in use and invented a different alias for Jaeger. Didn't think about the possibility of renaming the existing one :). Will do it now.

volmedo commented:
> Sorry, let me take back my approval :-) Just realized that it's still missing the serialization tests, to ensure backwards compatibility (which is probably why this is still WIP).

Yes, it's WIP because of that. I added three items to a todo list for the PR and I began with the last one instead of the first one :).

volmedo commented:
I added the unmarshalling code and the corresponding tests.

I implemented `encoding.UnmarshalText` for the type instead of `json.UnmarshalJSON` since we are dealing with a single `string` value. The [docs say](https://golang.org/pkg/encoding/json/?source=post_page---------------------------#Unmarshal) that `json.Unmarshal` uses `encoding.UnmarshalText` if `json.UnmarshalJSON` is not available.

Aside from the tests, I also did some manual tests applying several modified yaml files and everything worked as expected. BTW, is there any end to end test that involves applying malformed deployment yaml files?

I didn't include code for marshalling. Since the underlying type is `string`, marshalling comes for free. I wrote tests to check this and to add protection for custom marshalling code that could be added in the future.

I've also removed the `WIP` label from the PR as it should be ready now :)

jpkrohling commented on pkg/apis/jaegertracing/v1/deployment_strategy.go:
```diff
@@ -0,0 +1,49 @@
+// UnmarshalText implements encoding.TextUnmarshaler to ensure that JSON values in the
+// strategy field of JSON jaeger specs are interpreted in a case-insensitive manner
+func (ds *DeploymentStrategy) UnmarshalText(text []byte) error {
+	if ds == nil {
```
This is interesting. I'm probably missing something obvious, but how would this be triggered? Wouldn't a nil object fail with a nil reference before reaching this code? Can you come up with a test case for this?

volmedo replied:
It's not obvious. It is possible for the function to be called on a nil receiver when interfaces are involved (as is the case here).

I was digging into the docs and I had a look at how [json.RawMessage.UnmarshalJSON](https://golang.org/src/encoding/json/stream.go?s=6952:7005#L257) is implemented. The check for nil felt strange, so I did a bit of research and came up with this https://tour.golang.org/methods/12

Wonderful Go! :)

volmedo replied:
BTW, I guess this kind of check should be implemented in other types implementing interfaces for additional safety. `Options` come to my mind, implementing both `json.Marshaler` and `json.Unmarshaler` interfaces, but I'm sure there are several others.

jpkrohling replied:
TIL! I'm merging this one. Would you mind creating an issue to do the same nil check in the other relevant types?

volmedo replied:
Will do!

jpkrohling commented:
Thanks for your contribution!
